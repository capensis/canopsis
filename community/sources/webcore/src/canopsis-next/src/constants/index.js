export * from './recurrence-rule';
export * from './common';
export * from './entity';
export * from './alarm';
export * from './broadcast-message';
export * from './date';
export * from './dynamic-info';
export * from './event-filter';
export * from './pattern';
export * from './healthcheck';
export * from './idle-rule';
export * from './junit';
export * from './modal';
export * from './pbehavior';
export * from './permission';
export * from './remediation';
export * from './scenario';
export * from './snmp';
export * from './stats';
export * from './store';
export * from './widget';
export * from './kpi';
export * from './user';
export * from './service-weather';
export * from './map';
export * from './flowchart';
export * from './parameters';
export * from './declare-ticket';
export * from './link-rule';
export * from './external-data';
export * from './tag';
export * from './counter';
export * from './calendar';
export * from './icon';
export * from './availability';
