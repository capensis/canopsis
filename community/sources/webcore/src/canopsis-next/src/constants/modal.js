export const MODALS = {
  createEvent: 'create-event',
  createAckEvent: 'create-ack-event',
  confirmAckWithTicket: 'confirm-ack-with-ticket',
  createAssociateTicketEvent: 'create-associate-ticket-event',
  createCommentEvent: 'create-comment-event',
  createChangeStateEvent: 'create-change-state-event',
  createSnoozeEvent: 'create-snooze-event',
  variablesHelp: 'variables-help',
  createEntity: 'create-entity',
  createService: 'create-service',
  entityDependenciesList: 'entity-dependencies-list',
  createEntityInfo: 'create-entity-info',
  serviceEntities: 'service-entities',
  serviceDependencies: 'service-dependencies',
  createServicePauseEvent: 'create-service-pause-event',
  editLiveReporting: 'edit-live-reporting',
  infoPopupSetting: 'info-popup-setting',
  addInfoPopup: 'add-info-popup',
  confirmation: 'confirmation',
  confirmationPhrase: 'confirmation-phrase',
  clickOutsideConfirmation: 'click-outside-confirmation',
  createWidget: 'create-widget',
  selectWidgetTemplateType: 'select-widget-template-type',
  createWidgetTemplate: 'create-widget-template',
  createFilter: 'create-filter',
  alarmsList: 'alarms-list',
  createAlarmChart: 'create-alarm-chart',
  colorPicker: 'color-picker',
  textEditor: 'text-editor',
  textEditorWithTemplate: 'text-editor-with-template',
  textFieldEditor: 'text-field-editor',
  selectView: 'select-view',
  createView: 'create-view',
  shareView: 'share-view',
  createGroup: 'create-group',
  createUser: 'create-user',
  createRole: 'create-role',
  createBroadcastMessage: 'create-broadcast-message',
  createEventFilter: 'create-event-filter',
  createMetaAlarmRule: 'create-meta-alarm-rule',
  createPattern: 'create-pattern',
  eventFilterActions: 'event-filter-actions',
  createEventFilterAction: 'create-event-filter-action',
  filtersList: 'filters-list',
  createSnmpRule: 'create-snmp-rule',
  selectViewTab: 'select-view-tab',
  createDynamicInfo: 'create-dynamic-info',
  createDynamicInfoInformation: 'create-dynamic-info-information',
  importExportViews: 'import-groups-and-views',
  dynamicInfoTemplatesList: 'dynamic-info-templates-list',
  createDynamicInfoTemplate: 'create-dynamic-info-template',
  createPlaylist: 'create-playlist',
  managePlaylistTabs: 'manage-playlist-tabs',

  pbehaviorList: 'pbehavior-list',
  pbehaviorPlanning: 'pbehavior-planning',
  pbehaviorsCalendar: 'pbehaviors-calendar',
  pbehaviorRecurrenceRule: 'pbehavior-recurrence-rule',
  createRecurrenceRule: 'create-recurrence-rule',
  selectExceptionsLists: 'select-exceptions-lists',
  pbehaviorRecurrentChangesConfirmation: 'pbehavior-recurrent-changes-confirmation',
  createPbehavior: 'create-pbehavior',
  createPbehaviorType: 'create-pbehavior-type',
  createPbehaviorReason: 'create-pbehavior-reason',
  createPbehaviorException: 'create-pbehavior-exception',
  importPbehaviorException: 'import-pbehavior-exception',

  createManualMetaAlarm: 'create-manual-meta-alarm',
  removeAlarmsFromMetaAlarm: 'remove-alarms-from-meta-alarm',
  createRemediationInstruction: 'create-remediation-instruction',
  createRemediationConfiguration: 'create-remediation-configuration',
  createRemediationJob: 'create-remediation-job',
  createRemediationInstructionsFilter: 'create-remediation-instructions-filter',
  executeRemediationInstruction: 'execute-remediation-instruction',
  executeRemediationSimpleInstruction: 'execute-remediation-simple-instruction',
  remediationPatterns: 'remediation-patterns',
  pbehaviorPatterns: 'pbehavior-patterns',
  remediationInstructionApproval: 'remediation-instruction-approval',

  imageViewer: 'image-viewer',
  imagesViewer: 'images-viewer',
  rate: 'rate',
  createScenario: 'create-scenario',
  stateSetting: 'state-setting',
  testSuite: 'test-suite',
  createIdleRule: 'create-idle-rule',
  createAlarmStatusRule: 'create-alarm-status-rule',
  healthcheckEngine: 'healthcheck-engine',
  healthcheckEnginesChainReference: 'healthcheck-engines-chain-reference',
  info: 'info',
  createKpiFilter: 'create-kpi-filter',
  createMap: 'create-map',
  createGeoMap: 'create-geo-map',
  createFlowchartMap: 'create-flowchart-map',
  createMermaidMap: 'create-mermaid-map',
  createTreeOfDependenciesMap: 'create-tree-of-dependencies-map',

  createShareToken: 'create-share-token',

  createDeclareTicketEvent: 'create-declare-ticket-event',
  createDeclareTicketRule: 'create-declare-ticket-rule',
  executeDeclareTickets: 'execute-declare-tickets',

  createLinkRule: 'create-link-rule',
};
