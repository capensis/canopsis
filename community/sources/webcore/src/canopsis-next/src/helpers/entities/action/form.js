import { omit } from 'lodash';

import { ENTITIES_STATES, ACTION_TYPES, PATTERNS_FIELDS, OLD_PATTERNS_FIELDS } from '@/constants';

import { uid } from '@/helpers/uid';
import { durationToForm } from '@/helpers/date/duration';
import { getLocaleTimezone } from '@/helpers/date/date';

import { formToPbehavior, pbehaviorToForm, pbehaviorToRequest } from '../pbehavior/form';
import { requestToForm, formToRequest } from '../shared/request/form';
import { eventToAssociateTicketForm, formToAssociateTicketEvent } from '../associate-ticket/event/form';
import {
  declareTicketRuleWebhookDeclareTicketToForm,
  formToDeclareTicketRuleWebhookDeclareTicket,
} from '../declare-ticket/rule/form';
import { filterPatternsToForm, formFilterToPatterns } from '../filter/form';

/**
 * @typedef {
 *   "snooze" |
 *   "pbehavior" |
 *   "changestate" |
 *   "webhook" |
 *   "ack" |
 *   "ackremove" |
 *   "assocticket" |
 *   "cancel"
 * } ActionType
 */

/**
 * @typedef {Object} ActionForwardAuthorParameters
 * @property {boolean} [forward_author]
 * @property {string} [author]
 */

/**
 * @typedef {ActionForwardAuthorParameters} ActionDefaultParameters
 * @property {string} output
 */

/**
 * @typedef {ActionDefaultParameters} ActionSnoozeParameters
 * @property {Duration} duration
 */

/**
 * @typedef {ActionDefaultParameters} ActionChangeStateParameters
 * @property {number} state
 */

/**
 * @typedef {ActionDefaultParameters & AssociateTicketEvent} ActionAssocTicketParameters
 */

/**
 * @typedef {Object} ActionWebhookParameters
 * @property {Request} request
 * @property {?DeclareTicketRuleWebhookDeclareTicket} [declare_ticket]
 * @property {boolean} [forward_author]
 * @property {boolean} skip_for_child
 * @property {boolean} skip_for_instruction
 * @property {string} [author]
 */

/**
 * @typedef {ActionWebhookParameters} ActionWebhookFormParameters
 * @property {RequestForm} request
 * @property {DeclareTicketRuleWebhookDeclareTicketForm} declare_ticket
 */

/**
 * @typedef {
 *   Pbehavior |
 *   ActionDefaultParameters |
 *   ActionSnoozeParameters |
 *   ActionChangeStateParameters |
 *   ActionWebhookParameters |
 *   ActionAssocTicketParameters
 * } ActionParameters
 */

/**
 * @typedef {FilterPatterns} Action
 * @property {ActionType} type
 * @property {boolean} drop_scenario_if_not_matched
 * @property {boolean} emit_trigger
 * @property {string} comment
 * @property {ActionParameters} parameters
 */

/**
 * @typedef {
 *   PbehaviorForm |
 *   ActionDefaultParameters |
 *   ActionSnoozeParameters |
 *   ActionChangeStateParameters |
 *   ActionWebhookFormParameters |
 *   ActionAssocTicketParameters
 * } ActionFormParameters
 */

/**
 * @typedef {Action} ActionForm
 * @property {FilterPatternsForm} patterns
 * @property {Object.<ActionType, ActionFormParameters>} parameters
 */

/**
 * Check action type is pbehavior
 *
 * @param {ActionType} type
 * @return {boolean}
 */
export const isPbehaviorActionType = type => type === ACTION_TYPES.pbehavior;

/**
 * Check action type is webhook
 *
 * @param {ActionType} type
 * @return {boolean}
 */
export const isWebhookActionType = type => type === ACTION_TYPES.webhook;

/**
 * Check if action type is associate ticket
 *
 * @param {ActionType} type
 * @returns {boolean}
 */
export const isAssociateTicketActionType = type => type === ACTION_TYPES.assocticket;

/**
 * Convert action parameters to form
 *
 * @param {ActionForwardAuthorParameters | {}} [parameters = {}]
 * @returns {ActionForwardAuthorParameters}
 */
const defaultActionForwardAuthorToForm = (parameters = {}) => ({
  forward_author: parameters.forward_author ?? true,
  author: parameters.author ?? '',
});

/**
 * Convert action parameters to form
 *
 * @param {ActionDefaultParameters | {}} [parameters = {}]
 * @returns {ActionDefaultParameters}
 */
const defaultActionParametersToForm = (parameters = {}) => ({
  ...defaultActionForwardAuthorToForm(parameters),
  output: parameters.output ?? '',
});

/**
 * Convert action webhook parameters to form
 *
 * @param {ActionWebhookParameters} [parameters = {}]
 * @returns {ActionWebhookFormParameters}
 */
const webhookActionParametersToForm = (parameters = {}) => ({
  ...defaultActionForwardAuthorToForm(parameters),
  declare_ticket: declareTicketRuleWebhookDeclareTicketToForm(parameters.declare_ticket),
  request: requestToForm(parameters.request),
  skip_for_child: parameters.skip_for_child ?? false,
  skip_for_instruction: parameters.skip_for_instruction ?? false,
});

/**
 * Convert action snooze parameters to form
 *
 * @param {ActionSnoozeParameters | {}} [parameters = {}]
 * @returns {ActionSnoozeParameters}
 */
const snoozeActionParametersToForm = (parameters = {}) => ({
  ...defaultActionParametersToForm(parameters),
  duration: durationToForm(parameters.duration),
});

/**
 * Convert action snooze parameters to form
 *
 * @param {ActionChangeStateParameters | {}} [parameters = {}]
 * @returns {ActionChangeStateParameters}
 */
const changeStateActionParametersToForm = (parameters = {}) => ({
  ...defaultActionParametersToForm(parameters),
  state: parameters.state ?? ENTITIES_STATES.minor,
});

/**
 * Convert action assoc ticket parameters to form
 *
 * @param {ActionAssocTicketParameters | {}} [parameters = {}]
 * @returns {ActionAssocTicketParameters}
 */
const assocTicketActionParametersToForm = (parameters = {}) => ({
  ...defaultActionParametersToForm(parameters),
  ...omit(eventToAssociateTicketForm(parameters), ['ticket_comment']),
});

/**
 * Convert action pbehavior parameters to form
 *
 * @param {Pbehavior} [parameters = {}]
 * @param {string} [timezone = getLocaleTimezone()]
 * @returns {PbehaviorForm}
 */
const pbehaviorActionParametersToForm = (parameters = {}, timezone = getLocaleTimezone()) => {
  const pbehaviorForm = pbehaviorToForm(parameters, null, timezone);

  pbehaviorForm.start_on_trigger = !!parameters.start_on_trigger;
  pbehaviorForm.duration = durationToForm(parameters.duration);

  return pbehaviorForm;
};

/**
 * Prepare parameters for all action types
 *
 * @returns {Object.<ActionType, ActionFormParameters>}
 */
const prepareDefaultActionParameters = () => ({
  [ACTION_TYPES.snooze]: snoozeActionParametersToForm(),
  [ACTION_TYPES.pbehavior]: pbehaviorActionParametersToForm(),
  [ACTION_TYPES.changeState]: changeStateActionParametersToForm(),
  [ACTION_TYPES.ack]: defaultActionParametersToForm(),
  [ACTION_TYPES.ackremove]: defaultActionParametersToForm(),
  [ACTION_TYPES.assocticket]: assocTicketActionParametersToForm(),
  [ACTION_TYPES.cancel]: defaultActionParametersToForm(),
  [ACTION_TYPES.webhook]: webhookActionParametersToForm(),
});

/**
 * Convert action parameters to form
 *
 * @param {Action} action
 * @param {string} timezone
 * @returns {Object.<ActionType, ActionParameters>}
 */
export const actionParametersToForm = (action, timezone) => {
  const parameters = prepareDefaultActionParameters();

  if (!action.type || !action.parameters) {
    return parameters;
  }

  const parametersPreparers = {
    [ACTION_TYPES.snooze]: snoozeActionParametersToForm,
    [ACTION_TYPES.webhook]: webhookActionParametersToForm,
    [ACTION_TYPES.pbehavior]: pbehaviorActionParametersToForm,
    [ACTION_TYPES.assocticket]: assocTicketActionParametersToForm,
    [ACTION_TYPES.changeState]: changeStateActionParametersToForm,
  };

  const prepareParametersToFormFunction = parametersPreparers[action.type];

  parameters[action.type] = prepareParametersToFormFunction
    ? prepareParametersToFormFunction(action.parameters, timezone)
    : defaultActionParametersToForm({ ...action.parameters });

  return parameters;
};

/**
 * Convert action to form
 *
 * @param {Action} [action = {}]
 * @param {string} [timezone = getLocaleTimezone()]
 * @returns {ActionForm}
 */
export const actionToForm = (action = {}, timezone = getLocaleTimezone()) => ({
  type: action.type ?? ACTION_TYPES.snooze,
  key: uid(),
  parameters: actionParametersToForm(action, timezone),
  drop_scenario_if_not_matched: !!action.drop_scenario_if_not_matched,
  emit_trigger: !!action.emit_trigger,
  comment: action.comment ?? '',
  patterns: filterPatternsToForm(
    action,
    [PATTERNS_FIELDS.alarm, PATTERNS_FIELDS.entity],
    [OLD_PATTERNS_FIELDS.alarm, OLD_PATTERNS_FIELDS.entity],
  ),
});

/**
 * Convert pbehavior parameters to action
 *
 * @param {ActionWebhookFormParameters | {}} [parameters = {}]
 * @return {ActionWebhookParameters}
 */
export const formToWebhookActionParameters = (parameters = {}) => ({
  declare_ticket: formToDeclareTicketRuleWebhookDeclareTicket(parameters.declare_ticket),
  request: formToRequest(parameters.request),
  skip_for_child: parameters.skip_for_child,
  skip_for_instruction: parameters.skip_for_instruction,
});

/**
 * Convert pbehavior parameters to action
 *
 * @param {PbehaviorForm | {}} [parameters = {}]
 * @param {string} [timezone = getLocaleTimezone()]
 * @return {PbehaviorRequest}
 */
export const formToPbehaviorActionParameters = (parameters = {}, timezone = getLocaleTimezone()) => {
  const pbehavior = formToPbehavior(omit(parameters, ['start_on_trigger', 'duration']), timezone);

  if (parameters.start_on_trigger) {
    pbehavior.start_on_trigger = parameters.start_on_trigger;
    pbehavior.duration = parameters.duration;
  }

  return pbehaviorToRequest(pbehavior);
};

/**
 * Convert form to action parameters
 *
 * @param {ActionForm} form
 * @param {string} [timezone]
 * @returns {ActionParameters}
 */
const formToActionParameters = (form, timezone) => {
  const {
    forward_author: forwardAuthor,
    author,
    ...parametersByCurrentType
  } = form.parameters[form.type];

  const parametersPreparers = {
    [ACTION_TYPES.webhook]: formToWebhookActionParameters,
    [ACTION_TYPES.pbehavior]: formToPbehaviorActionParameters,
    [ACTION_TYPES.assocticket]: formToAssociateTicketEvent,
  };

  const prepareParametersToAction = parametersPreparers[form.type];
  const parameters = prepareParametersToAction
    ? prepareParametersToAction(parametersByCurrentType, timezone)
    : parametersByCurrentType;

  if (!isPbehaviorActionType(form.type)) {
    parameters.forward_author = forwardAuthor;

    if (!forwardAuthor) {
      parameters.author = author;
    }
  }

  return parameters;
};

/**
 * Convert form to action
 *
 * @param {ActionForm} form
 * @param {string} [timezone]
 * @returns {Action}
 */
export const formToAction = (form, timezone) => ({
  ...omit(form, ['key', 'patterns']),
  ...formFilterToPatterns(form.patterns, [PATTERNS_FIELDS.alarm, PATTERNS_FIELDS.entity]),
  parameters: formToActionParameters(form, timezone),
});
