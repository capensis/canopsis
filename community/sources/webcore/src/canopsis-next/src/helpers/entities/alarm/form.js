/**
 * @typedef {Object} AlarmOldLink
 * @property {string} label
 * @property {string} link
 */

/**
 * @typedef {Object} AlarmLink
 * @property {string} icon_name
 * @property {string} label
 * @property {string} url
 * @property {string} [rule_id]
 * @property {LinkRuleAction} [action]
 */

/**
 * @typedef {Object<string, AlarmLink[] | AlarmOldLink[]>} AlarmLinks
 */

/**
 * @typedef {Object} AlarmEvent
 * @property {number} val
 * @property {string} a
 * @property {number} t
 * @property {string} _t
 * @property {string} m
 */

/**
 * @typedef {Object} Correlation
 * @property {number} total
 * @property {Alarm} data
 */

/**
 * @typedef {Object} AlarmRule
 * @property {string} id
 * @property {string} name
 */

/**
 * @typedef {Object} AlarmAssignedInstructionExecution
 * @property {string} _id
 * @property {number} status
 */

/**
 * @typedef {Object} AlarmAssignedInstruction
 * @property {string} _id
 * @property {string} name
 * @property {?AlarmAssignedInstructionExecution} execution
 */

/**
 * @typedef {Object} AlarmValuePbehaviorInfo
 * @property {string} canonical_type
 * @property {string} icon_name
 * @property {string} id
 * @property {string} name
 * @property {string} reason
 * @property {string} type
 * @property {string} type_name
 * @property {number} timestamp
 */

/**
 * @typedef {Object} AlarmValue
 * @property {string[]} long_output_history
 * @property {number} last_event_date
 * @property {string} connector_name
 * @property {string} initial_long_output
 * @property {string} output
 * @property {string} initial_output
 * @property {string[]} children
 * @property {string[]} filtered_children
 * @property {Object} extra
 * @property {number} last_update_date
 * @property {number} resolved
 * @property {string} resource
 * @property {number} creation_date
 * @property {string} display_name
 * @property {number} total_state_changes
 * @property {string[]} tags
 * @property {string} long_output
 * @property {string} component
 * @property {string} connector
 * @property {Object} infos_rule_version
 * @property {InfosObject} infos
 * @property {string[]} parents
 * @property {AlarmEvent} ack
 * @property {AlarmEvent} state
 * @property {AlarmEvent} status
 * @property {AlarmEvent} ticket
 * @property {AlarmEvent} snooze
 * @property {AlarmEvent} canceled
 * @property {AlarmEvent} last_comment
 * @property {AlarmEvent[]} [steps]
 * @property {AlarmValuePbehaviorInfo} [pbehavior_info]
 */

/**
 * @typedef {Pbehavior} AlarmPbehavior
 * @property {Comment} last_comment
 */

/**
 * @typedef {Object} Alarm
 * @property {string} _id
 * @property {Entity} entity
 * @property {boolean} is_meta_alarm
 * @property {AlarmAssignedInstruction[]} [assigned_instructions]
 * @property {number} [instruction_execution_icon]
 * @property {string[]} running_manual_instructions
 * @property {string[]} running_auto_instructions
 * @property {string[]} failed_manual_instructions
 * @property {string[]} failed_auto_instructions
 * @property {string[]} successful_manual_instructions
 * @property {string[]} successful_auto_instructions
 * @property {boolean} [children_instructions]
 * @property {InfosObject} infos
 * @property {AlarmRule} rule
 * @property {Correlation} consequences
 * @property {Correlation} causes
 * @property {AlarmPbehavior} pbehavior
 * @property {AlarmLinks} links
 * @property {number} t
 * @property {AlarmValue} v
 */

import { ENTITIES_STATES, ENTITIES_STATUSES } from '@/constants';

/**
 * Checks if alarm is resolved
 * @param alarm - alarm entity
 * @returns {boolean}
 */
export const isClosedAlarmStatus = alarm => [ENTITIES_STATUSES.closed, ENTITIES_STATUSES.cancelled]
  .includes(alarm.v.status.val);

/**
 * Checks if alarm is resolved
 *
 * @param alarm - alarm entity
 * @returns {boolean}
 */
export const isResolvedAlarm = alarm => !!alarm.v.resolved;

/**
 * Checks if alarm have critical state
 *
 * @param alarm - alarm entity
 * @returns {boolean}
 */
export const isAlarmStateNotOk = alarm => ENTITIES_STATES.ok !== alarm.v.state.val;
