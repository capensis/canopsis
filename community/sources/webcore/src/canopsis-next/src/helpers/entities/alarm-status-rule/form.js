import { omit } from 'lodash';

import { PATTERNS_FIELDS, TIME_UNITS } from '@/constants';

import { durationToForm } from '@/helpers/date/duration';
import { filterPatternsToForm, formFilterToPatterns } from '@/helpers/entities/filter/form';

/**
 * @typedef {FilterPatterns} AlarmStatusRule
 * @property {string} name
 * @property {string} description
 * @property {Duration} duration
 * @property {number} priority
 * @property {number} [freq_limit]
 */

/**
 * @typedef {Object} AlarmStatusRuleForm
 * @property {string} name
 * @property {string} description
 * @property {Duration} duration
 * @property {number} priority
 * @property {FilterPatternsForm} patterns
 */

/**
 * Convert alarm status rule object to form compatible object
 *
 * @param {AlarmStatusRule | {}} [rule = {}]
 * @param {boolean} [flapping = false]
 * @return {AlarmStatusRuleForm}
 */
export const alarmStatusRuleToForm = (rule = {}, flapping = false) => {
  const form = {
    name: rule.name ?? '',
    duration: rule.duration
      ? durationToForm(rule.duration)
      : { value: 1, unit: TIME_UNITS.minute },
    priority: rule.priority,
    description: rule.description ?? '',
    patterns: filterPatternsToForm(rule, [PATTERNS_FIELDS.alarm, PATTERNS_FIELDS.entity]),
  };

  if (flapping) {
    form.freq_limit = rule.freq_limit ?? 1;
  }

  return form;
};

/**
 * Convert form compatible object to alarm status rule object
 *
 * @param {AlarmStatusRuleForm} form
 * @return {AlarmStatusRule}
 */
export const formToAlarmStatusRule = form => ({
  ...omit(form, ['patterns']),
  ...formFilterToPatterns(form.patterns),
});
