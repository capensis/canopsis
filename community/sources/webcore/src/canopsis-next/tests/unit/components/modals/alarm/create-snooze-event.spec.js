import Faker from 'faker';

import { flushPromises, generateRenderer, generateShallowRenderer } from '@unit/utils/vue';
import { mockModals, mockPopups } from '@unit/utils/mock-hooks';
import { createButtonStub } from '@unit/stubs/button';
import { createFormStub } from '@unit/stubs/form';
import { createModalWrapperStub } from '@unit/stubs/modal';
import ClickOutside from '@/services/click-outside';
import { TIME_UNITS } from '@/constants';

import CreateSnoozeEvent from '@/components/modals/alarm/create-snooze-event.vue';

const stubs = {
  'modal-wrapper': createModalWrapperStub('modal-wrapper'),
  'snooze-event-form': true,
  'v-btn': createButtonStub('v-btn'),
  'v-form': createFormStub('v-form'),
};

const snapshotStubs = {
  'modal-wrapper': createModalWrapperStub('modal-wrapper'),
  'snooze-event-form': true,
};

const selectButtons = wrapper => wrapper.findAll('button.v-btn');
const selectSubmitButton = wrapper => selectButtons(wrapper).at(1);
const selectCancelButton = wrapper => selectButtons(wrapper).at(0);
const selectSnoozeEventForm = wrapper => wrapper.find('snooze-event-form-stub');

describe('create-snooze-event', () => {
  const $modals = mockModals();
  const $popups = mockPopups();

  const alarm = {
    _id: Faker.datatype.string(),
    v: {
      connector: Faker.datatype.string(),
      connector_name: Faker.datatype.string(),
      component: Faker.datatype.string(),
      resource: Faker.datatype.string(),
      state: {
        val: Faker.datatype.number(),
      },
      status: {
        val: Faker.datatype.number(),
      },
    },
    entity: {
      type: Faker.datatype.number(),
    },
  };
  const items = [alarm];
  const snoozeEventData = {
    duration: {
      unit: TIME_UNITS.minute,
      value: 1,
    },
    comment: '',
  };
  const config = { items };

  const factory = generateShallowRenderer(CreateSnoozeEvent, {
    stubs,
    attachTo: document.body,
    propsData: {
      modal: {
        config: {},
      },
    },
    parentComponent: {
      provide: {
        $clickOutside: new ClickOutside(),
      },
    },
  });
  const snapshotFactory = generateRenderer(CreateSnoozeEvent, {
    stubs: snapshotStubs,
    propsData: {
      modal: {
        config: {},
      },
    },
    parentComponent: {
      provide: {
        $clickOutside: new ClickOutside(),
      },
    },
  });

  const timestamp = 1386435600000;

  beforeAll(() => jest.useFakeTimers({ now: timestamp }));

  test('Default parameters applied to form', () => {
    const wrapper = factory({
      mocks: {
        $modals,
      },
    });

    const snoozeEventForm = selectSnoozeEventForm(wrapper);

    expect(snoozeEventForm.vm.form).toEqual({
      comment: '',
      duration: {
        unit: TIME_UNITS.minute,
        value: 1,
      },
    });
  });

  test('Form submitted after trigger submit button', async () => {
    const action = jest.fn();

    const wrapper = factory({
      propsData: {
        modal: {
          config: {
            items,
            action,
          },
        },
      },
      mocks: {
        $modals,
      },
    });

    const submitButton = selectSubmitButton(wrapper);

    submitButton.trigger('click');

    await flushPromises(true);

    expect(action).toBeCalledWith(snoozeEventData);
    expect($modals.hide).toBeCalledWith();
  });

  test('Form didn\'t submitted after trigger submit button with error', async () => {
    const action = jest.fn();
    const wrapper = factory({
      propsData: {
        modal: {
          config: {
            ...config,
            action,
          },
        },
      },
      mocks: {
        $modals,
      },
    });

    const snoozeEventForm = selectSnoozeEventForm(wrapper);

    const validator = wrapper.getValidator();

    validator.attach({
      name: 'name',
      rules: 'required:true',
      getter: () => false,
      context: () => snoozeEventForm.vm,
      vm: snoozeEventForm.vm,
    });

    const submitButton = selectSubmitButton(wrapper);

    submitButton.trigger('click');

    await flushPromises(true);

    expect(action).not.toBeCalled();

    validator.detach('name');
  });

  test('Errors added after trigger submit button with action errors', async () => {
    const action = jest.fn();
    const formErrors = {
      duration: 'Ticket error',
      comment: 'Comment error',
    };
    action.mockRejectedValueOnce({ ...formErrors, unavailableField: 'Error' });
    const wrapper = factory({
      propsData: {
        modal: {
          config: {
            ...config,
            action,
          },
        },
      },
      mocks: {
        $modals,
      },
    });

    const submitButton = selectSubmitButton(wrapper);

    submitButton.trigger('click');

    await flushPromises(true);

    const addedErrors = wrapper.getValidatorErrorsObject();

    expect(formErrors).toEqual(addedErrors);
    expect(action).toBeCalledWith(snoozeEventData);
    expect($modals.hide).not.toBeCalledWith();

    action.mockClear();
  });

  test('Error popup showed after trigger submit button with action errors', async () => {
    const action = jest.fn();
    const consoleErrorSpy = jest.spyOn(console, 'error').mockImplementation();
    const errors = {
      unavailableField: 'Error',
      anotherUnavailableField: 'Second error',
    };
    action.mockRejectedValueOnce(errors);

    const wrapper = factory({
      propsData: {
        modal: {
          config: {
            ...config,
            action,
          },
        },
      },
      mocks: {
        $modals,
        $popups,
      },
    });

    const submitButton = selectSubmitButton(wrapper);

    submitButton.trigger('click');

    await flushPromises(true);

    expect(consoleErrorSpy).toBeCalledWith(errors);
    expect($popups.error).toBeCalledWith({
      text: `${errors.unavailableField}\n${errors.anotherUnavailableField}`,
    });
    expect(action).toBeCalledWith(snoozeEventData);
    expect($modals.hide).not.toBeCalledWith();

    consoleErrorSpy.mockClear();
    action.mockClear();
  });

  test('Modal submitted with correct data after trigger form', async () => {
    const action = jest.fn();
    const wrapper = factory({
      propsData: {
        modal: {
          config: {
            ...config,
            action,
          },
        },
      },
      mocks: {
        $modals,
      },
    });

    const snoozeEventForm = selectSnoozeEventForm(wrapper);

    const newForm = {
      duration: {
        unit: TIME_UNITS.hour,
        value: 2,
      },
      comment: 'comment',
    };

    snoozeEventForm.vm.$emit('input', newForm);

    const submitButton = selectSubmitButton(wrapper);

    submitButton.trigger('click');

    await flushPromises(true);

    expect(action).toBeCalledWith(newForm);
    expect($modals.hide).toBeCalled();
  });

  test('Modal hidden after trigger cancel button', async () => {
    const wrapper = factory({
      propsData: {
        modal: {
          config,
        },
      },
      mocks: {
        $modals,
      },
    });

    const cancelButton = selectCancelButton(wrapper);

    cancelButton.trigger('click');

    await flushPromises(true);

    expect($modals.hide).toBeCalled();
  });

  test('Renders `create-snooze-event` with empty modal', () => {
    const wrapper = snapshotFactory({
      propsData: {
        modal: {
          config,
        },
      },
      mocks: {
        $modals,
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  test('Renders `create-snooze-event` with config data', () => {
    const wrapper = snapshotFactory({
      propsData: {
        modal: {
          config: {
            items,
            isNoteRequired: true,
          },
        },
      },
      mocks: {
        $modals,
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });
});
