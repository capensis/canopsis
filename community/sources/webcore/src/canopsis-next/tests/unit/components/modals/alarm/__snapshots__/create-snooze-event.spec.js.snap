// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create-snooze-event Renders \`create-snooze-event\` with config data 1`] = `
<form
  class="v-form"
  novalidate="novalidate"
>
  <div
    class="modal-wrapper"
    close=""
  >
    <div
      class="title"
    >
      <span>
        Snooze
      </span>
    </div>
     
    <div
      class="text"
    >
      <snooze-event-form-stub
        form="[object Object]"
        isnoterequired="true"
      />
    </div>
     
    <div
      class="actions"
    >
      <button
        class="v-btn v-btn--text theme--light v-size--default"
        type="button"
      >
        <span
          class="v-btn__content"
        >
          
        Cancel
      
        </span>
      </button>
       
      <button
        class="primary v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default"
        type="submit"
      >
        <span
          class="v-btn__content"
        >
          
        Save changes
      
        </span>
      </button>
    </div>
  </div>
</form>
`;

exports[`create-snooze-event Renders \`create-snooze-event\` with empty modal 1`] = `
<form
  class="v-form"
  novalidate="novalidate"
>
  <div
    class="modal-wrapper"
    close=""
  >
    <div
      class="title"
    >
      <span>
        Snooze
      </span>
    </div>
     
    <div
      class="text"
    >
      <snooze-event-form-stub
        form="[object Object]"
      />
    </div>
     
    <div
      class="actions"
    >
      <button
        class="v-btn v-btn--text theme--light v-size--default"
        type="button"
      >
        <span
          class="v-btn__content"
        >
          
        Cancel
      
        </span>
      </button>
       
      <button
        class="primary v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default"
        type="submit"
      >
        <span
          class="v-btn__content"
        >
          
        Save changes
      
        </span>
      </button>
    </div>
  </div>
</form>
`;
